### PR TITLE
url: fix various type error

### PR DESCRIPTION
**Pull request by Exirel**
### Description

Making mypy happy with `url.py`.

### Checklist

- [x] I have read [CONTRIBUTING.md](https://github.com/sopel-irc/sopel/blob/master/CONTRIBUTING.md)
- [x] I can and do license this contribution under the EFLv2
- [x] No issues are reported by `make qa` (runs `make quality` and `make test`)
- [x] I have tested the functionality of the things this change touches

### CODE REVIEW

File: sopel/modules/url.py [hunk 1]
```diff
@@ -360,6 +367,11 @@ def process_urls(
         # FIXME: This does nothing when an attacker knows how to host a 302
         # FIXME: This whole concept has a TOCTOU issue
         if not bot.config.url.enable_private_resolution:
+            if not parsed_url.hostname:
```
Comment (half-duplex) [inline sopel/modules/url.py:370]: I'm guessing this is an unrelated bugfix?

**Review by half-duplex — APPROVED** [review]

File: sopel/modules/url.py (hunk 1 above)
Comment (Exirel) [reply sopel/modules/url.py:370]: Nope, `hostname` is `Optional[str]`, because `file:///path` will have `None` for the hostname, and yes this is a valid URL. However, `ip` requires `str`, so mypy legitimately complains about it.
